File: solutions/clojure/change/1/src/change.clj
;; respectfully dedicated to Shree DR.MDD
(ns change)

(defn build-lookup-table [goal pool]
  (let [options (sort pool)]
    (loop [i 1, ledger {0 []}]
      (if (= i (inc goal))
        ledger
        (->> (map #(vector % (- i %)) options)
             (filter (fn [[_ rem]] (>= rem 0)))
             (map #(conj (get ledger (second %)) (first %)))
             (filter vector?)
             (sort-by count)
             first
             (assoc ledger i)
             (recur (inc i)))))))

(defn issue [target coins]
  (cond
    (neg? target) (throw (IllegalArgumentException. "target can't be negative"))
    :else
    (let [chart (build-lookup-table target coins)
          answer (chart target)]
      (if answer
        (sort answer)
        (throw (IllegalArgumentException. "can't make target with given coins"))))))

File: solutions/ocaml/change/1/change.ml
let rec gcd x = function 0 -> x | y -> gcd y (x mod y)
let rec try_change change = function
  | 0, _, _ -> Some (List.sort Int.compare change)
  | _, 0, _ | _, _, [] -> None
  | target, count, coin :: coins ->
      match try_change change (target, count, coins) with
      | Some change -> Some change
      | None -> 
          try_change 
            (coin :: change) 
            (target - coin, count - 1, coin :: coins)
let make_change ~target ~coins =
  let open Result in
  if target < 0
  then error "target can't be negative"
  else if target mod List.fold_left gcd 0 coins > 0
  then error "can't make target with given coins"
  else
    let rec loop count =
      match try_change [] (target, count, coins) with
      | Some change -> ok change
      | None -> loop (count + 1)
    in
    loop 0